### ISSUE TITLE
Improve the "Failed to build gem native extension." error message.

### ISSUE THREAD

**Issue by voxik** [ruby/rubygems#5307]
Doing a lot of rebuilds on remote builders, this error message is quite unhelpful, misleading and annoying:

~~~
$ CONFIGURE_ARGS='--with-cflags='\''-O2 -flto=auto -ffat-lto-objects -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1  -m64 -mcpu=power8 -mtune=power8 -fasynchronous-unwind-tables -fstack-clash-protection'\'' --with-cxxflags='\''-O2 -flto=auto -ffat-lto-objects -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1  -m64 -mcpu=power8 -mtune=power8 -fasynchronous-unwind-tables -fstack-clash-protection'\'' --with-ldflags='\''-Wl,-z,relro -Wl,--as-needed  -Wl,-z,now -specs=/usr/lib/rpm/redhat/redhat-hardened-ld -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1  -Wl,--build-id=sha1 -Wl,-dT,/builddir/build/BUILD/.package_note-rubygem-hpricot-0.8.6-34.fc36.ppc64le.ld'\'' '
$ gem install -V --local --build-root . --force --document=ri,rdoc hpricot-0.8.6.gem

... snip ...

Building native extensions. This could take a while...
current directory: /builddir/build/BUILD/rubygem-hpricot-0.8.6/usr/share/gems/gems/hpricot-0.8.6/ext/hpricot_scan
["/usr/bin/ruby", "-I", "/usr/share/rubygems", "-r", "./siteconf20220125-118021-hpigkj.rb", "extconf.rb"]
ERROR:  Error installing hpricot-0.8.6.gem:
	ERROR: Failed to build gem native extension.
    current directory: /builddir/build/BUILD/rubygem-hpricot-0.8.6/usr/share/gems/gems/hpricot-0.8.6/ext/hpricot_scan
/usr/bin/ruby -I /usr/share/rubygems -r ./siteconf20220125-118021-hpigkj.rb extconf.rb
Building has failed. See above output for more information on the failure.
To see why this extension failed to compile, please check the mkmf.log which can be found here:
  /builddir/build/BUILD/rubygem-hpricot-0.8.6/usr/lib64/gems/ruby/hpricot-0.8.6/mkmf.log
extconf failed, exit code 1
Gem files will remain installed in /builddir/build/BUILD/rubygem-hpricot-0.8.6/usr/share/gems/gems/hpricot-0.8.6 for inspection.
Results logged to /builddir/build/BUILD/rubygem-hpricot-0.8.6/usr/lib64/gems/ruby/hpricot-0.8.6/gem_make.out
checking for main() in -lc... *** extconf.rb failed ***
Could not create Makefile due to some reason, probably lack of necessary
libraries and/or headers.  Check the mkmf.log file for more details.  You may
need configuration options.
Provided configuration options:
	--with-opt-dir
	--without-opt-dir
	--with-opt-include
	--without-opt-include=${opt-dir}/include
	--with-opt-lib
	--without-opt-lib=${opt-dir}/lib64
	--with-make-prog
	--without-make-prog
	--srcdir=.
	--curdir
	--ruby=/usr/bin/$(RUBY_BASE_NAME)
	--with-hpricot_scan-dir
	--without-hpricot_scan-dir
	--with-hpricot_scan-include
	--without-hpricot_scan-include=${hpricot_scan-dir}/include
	--with-hpricot_scan-lib
	--without-hpricot_scan-lib=${hpricot_scan-dir}/lib64
	--with-c-dir
	--without-c-dir
	--with-c-include
	--without-c-include=${c-dir}/include
	--with-c-lib
	--without-c-lib=${c-dir}/lib64
	--with-clib
	--without-clib
/usr/share/ruby/mkmf.rb:471:in `try_do': The compiler failed to generate an executable file. (RuntimeError)
You have to install development tools first.
	from /usr/share/ruby/mkmf.rb:564:in `try_link0'
	from /usr/share/ruby/mkmf.rb:582:in `try_link'
	from /usr/share/ruby/mkmf.rb:794:in `try_func'
	from /usr/share/ruby/mkmf.rb:1029:in `block in have_library'
	from /usr/share/ruby/mkmf.rb:971:in `block in checking_for'
	from /usr/share/ruby/mkmf.rb:361:in `block (2 levels) in postpone'
	from /usr/share/ruby/mkmf.rb:331:in `open'
	from /usr/share/ruby/mkmf.rb:361:in `block in postpone'
	from /usr/share/ruby/mkmf.rb:331:in `open'
	from /usr/share/ruby/mkmf.rb:357:in `postpone'
	from /usr/share/ruby/mkmf.rb:970:in `checking_for'
	from /usr/share/ruby/mkmf.rb:1024:in `have_library'
	from extconf.rb:7:in `<main>'
~~~

While it tries to be useful, it provides just all the wrong information.

From the end, it first says "You have to install development tools first.". That is likely not true, the development tools are there and the chances that something would significantly changed from previous build are small.

Then it provides configuration options, but I don't remember where I'd need to use configuration options. On Fedora build system, the probability that something like this would be the solution is close to zero.

Lastly, it suggests "Check the mkmf.log file for more details.". However, the build done on remote builder, I don't have access to the file. It would be much better if the content was printed there instead.

TL;DR

The smallest improvement would be to include the mkmf.log for failed builds. It might be sometimes longer, but it is never awfully long in my experience.

Should it be too long, the last section of mkmf.log relevant to `checking for main() in -lc...` would be certainly enough.

**Comment by deivid-rodriguez** [comment]
I believe this should be reported to ruby-core, since most of this logic lives at `mkmf.rb`. I actually have a WIP branch with some improvements along these lines, I'm happy to share it and propose it to ruby-core.

**Comment by simi** [comment]
> Lastly, it suggests "Check the mkmf.log file for more details.". However, the build done on remote builder, I don't have access to the file. It would be much better if the content was printed there instead.

:thinking: It think for `--verbose` flag we can output `mkmf.log` as well.

**Comment by voxik** [comment]
I have just looked at the code and I think this is completely in RubyGems hands, because RubyGems executes the extconf.rb:

https://github.com/rubygems/rubygems/blob/62b9f4af3e306a44ee0cc35d3004107f75f78cc6/lib/rubygems/ext/ext_conf_builder.rb#L46-L59

Of course something like this could be moved entirely into mkmf.

BTW the `begin`/`end` block around this code is nice historical relict :stuck_out_tongue_winking_eye:

**Comment by deivid-rodriguez** [comment]
Yes, that specific part yes, that's why I didn't close the issue. But the other things you mentioned (configuration options, you have to install development tools) are printed directly by mkmf.rb.

**Comment by rubyFeedback** [comment]
I agree that this should be improved - whether gem itself or ruby-core is, IMO, even though relevant, still secondary. mame improved lots of things in regards to error-reporting in ruby, and I am kind of used to good, useful error messages, so when gem finds an error I think it would be very helpful to notify the user what this is all about.

**Comment by casperisfine** [comment]
So I keep running into this problem on various CI system where getting the log file is either complicated or impossible.

I also very frequently see people report compilation failures with no relevant information at all since everything is in the log file.

A "verbose" option would be nice, and I understand that printing everything might not be desirable given that some extensions might have gigantic logs.

However I think the relevant information is very likely to be in the last few lines, so I wonder if we could just print the last 10 lines on failure by default or something.

What do you think?

**Comment by deivid-rodriguez** [comment]
That sounds good to me.

**Comment by simi** [comment]
@casperisfine do you have any details for your failures? I tried to reproduce C compilation time error, and it reports back all the details.

```ruby
[retro@retro  tmp]❤ gem install win32-api
Building native extensions. This could take a while...
ERROR:  Error installing win32-api:
        ERROR: Failed to build gem native extension.

    current directory: /home/retro/.gem/ruby/3.0.4/gems/win32-api-1.10.1/ext
/home/retro/.rubies/ruby-3.0.4/bin/ruby -I /home/retro/.rubies/ruby-3.0.4/lib/ruby/3.0.0 -r ./siteconf20221004-181841-gseiwu.rb extconf.rb
checking for strncpy_s()... no
creating Makefile

current directory: /home/retro/.gem/ruby/3.0.4/gems/win32-api-1.10.1/ext
make DESTDIR\= clean

current directory: /home/retro/.gem/ruby/3.0.4/gems/win32-api-1.10.1/ext
make DESTDIR\=
compiling win32/api.c
win32/api.c:2:10: fatal error: windows.h: Adresář nebo soubor neexistuje
    2 | #include <windows.h>
      |          ^~~~~~~~~~~
compilation terminated.
make: *** [Makefile:245: api.o] Chyba 1

make failed, exit code 2

Gem files will remain installed in /home/retro/.gem/ruby/3.0.4/gems/win32-api-1.10.1 for inspection.
Results logged to /home/retro/.gem/ruby/3.0.4/extensions/x86_64-linux/3.0.0-static/win32-api-1.10.1/gem_make.out
```

```
[retro@retro  tmp]❤ cat /home/retro/.gem/ruby/3.0.4/extensions/x86_64-linux/3.0.0-static/win32-api-1.10.1/gem_make.out
current directory: /home/retro/.gem/ruby/3.0.4/gems/win32-api-1.10.1/ext
/home/retro/.rubies/ruby-3.0.4/bin/ruby -I /home/retro/.rubies/ruby-3.0.4/lib/ruby/3.0.0 -r ./siteconf20221004-181841-gseiwu.rb extconf.rb
checking for strncpy_s()... no
creating Makefile

current directory: /home/retro/.gem/ruby/3.0.4/gems/win32-api-1.10.1/ext
make DESTDIR\= clean

current directory: /home/retro/.gem/ruby/3.0.4/gems/win32-api-1.10.1/ext
make DESTDIR\=
compiling win32/api.c
win32/api.c:2:10: fatal error: windows.h: Adresář nebo soubor neexistuje
    2 | #include <windows.h>
      |          ^~~~~~~~~~~
compilation terminated.
make: *** [Makefile:245: api.o] Chyba 1

make failed, exit code 2
```

**Comment by casperisfine** [comment]
🤦 I think you're right, not sure why I didn't see it maybe it changed recently.

The last occurence that prompted me to comment happens to be public: https://buildkite.com/rails/rails/builds/90069#0183a201-ac81-43d6-8476-bc7568f8d63a/2727-3141

```
Installing tailwindcss-rails 2.0.14 (x86_64-linux)
Gem::Ext::BuildError: ERROR: Failed to build gem native extension.
 
current directory:
/usr/local/lib/ruby/gems/3.2.0+2/gems/sqlite3-1.5.2/ext/sqlite3
/usr/local/bin/ruby -I /usr/local/lib/ruby/3.2.0+2 extconf.rb
Building sqlite3-ruby using packaged sqlite3.
Extracting sqlite-autoconf-3390400.tar.gz into
tmp/x86_64-linux-gnu/ports/sqlite3/3.39.4... OK
Running 'configure' for sqlite3 3.39.4... OK
Running 'compile' for sqlite3 3.39.4... OK
Running 'install' for sqlite3 3.39.4... OK
Activating sqlite3 3.39.4 (from
/usr/local/lib/ruby/gems/3.2.0+2/gems/sqlite3-1.5.2/ports/x86_64-linux-gnu/sqlite3/3.39.4)...
 
Could not configure the build properly. Please install either the `pkg-config`
utility or the `pkg-config` rubygem.
 
*** extconf.rb failed ***
Could not create Makefile due to some reason, probably lack of necessary
libraries and/or headers.  Check the mkmf.log file for more details.  You may
need configuration options.
 
Provided configuration options:
	--with-opt-dir
	--without-opt-dir
	--with-opt-include
	--without-opt-include=${opt-dir}/include
	--with-opt-lib
	--without-opt-lib=${opt-dir}/lib
	--with-make-prog
	--without-make-prog
	--srcdir=.
	--curdir
	--ruby=/usr/local/bin/$(RUBY_BASE_NAME)
	--help
	--download-dependencies
	--with-sqlcipher
	--without-sqlcipher
	--with-sqlcipher-dir
	--without-sqlcipher-dir
	--with-sqlcipher-include
	--without-sqlcipher-include
	--with-sqlcipher-lib
	--without-sqlcipher-lib
	--enable-system-libraries
	--disable-system-libraries
	--with-sqlcipher
	--without-sqlcipher
	--with-sqlcipher-dir
	--without-sqlcipher-dir
	--with-sqlcipher-include
	--without-sqlcipher-include
	--with-sqlcipher-lib
	--without-sqlcipher-lib
--with-/usr/local/lib/ruby/gems/3.2.0+2/gems/sqlite3-1.5.2/ports/x86_64-linux-gnu/sqlite3/3.39.4/lib/pkgconfig/sqlite3.pc-dir
--without-/usr/local/lib/ruby/gems/3.2.0+2/gems/sqlite3-1.5.2/ports/x86_64-linux-gnu/sqlite3/3.39.4/lib/pkgconfig/sqlite3.pc-dir
--with-/usr/local/lib/ruby/gems/3.2.0+2/gems/sqlite3-1.5.2/ports/x86_64-linux-gnu/sqlite3/3.39.4/lib/pkgconfig/sqlite3.pc-include
--without-/usr/local/lib/ruby/gems/3.2.0+2/gems/sqlite3-1.5.2/ports/x86_64-linux-gnu/sqlite3/3.39.4/lib/pkgconfig/sqlite3.pc-include=${/usr/local/lib/ruby/gems/3.2.0+2/gems/sqlite3-1.5.2/ports/x86_64-linux-gnu/sqlite3/3.39.4/lib/pkgconfig/sqlite3.pc-dir}/include
--with-/usr/local/lib/ruby/gems/3.2.0+2/gems/sqlite3-1.5.2/ports/x86_64-linux-gnu/sqlite3/3.39.4/lib/pkgconfig/sqlite3.pc-lib
--without-/usr/local/lib/ruby/gems/3.2.0+2/gems/sqlite3-1.5.2/ports/x86_64-linux-gnu/sqlite3/3.39.4/lib/pkgconfig/sqlite3.pc-lib=${/usr/local/lib/ruby/gems/3.2.0+2/gems/sqlite3-1.5.2/ports/x86_64-linux-gnu/sqlite3/3.39.4/lib/pkgconfig/sqlite3.pc-dir}/lib
--with-/usr/local/lib/ruby/gems/3.2.0+2/gems/sqlite3-1.5.2/ports/x86_64-linux-gnu/sqlite3/3.39.4/lib/pkgconfig/sqlite3.pc-config
--without-/usr/local/lib/ruby/gems/3.2.0+2/gems/sqlite3-1.5.2/ports/x86_64-linux-gnu/sqlite3/3.39.4/lib/pkgconfig/sqlite3.pc-config
	--with-pkg-config
	--without-pkg-config
 
To see why this extension failed to compile, please check the mkmf.log which can
be found here:
 
/usr/local/lib/ruby/gems/3.2.0+2/extensions/x86_64-linux/3.2.0+2/sqlite3-1.5.2/mkmf.log
 
extconf failed, exit code 1
 
Gem files will remain installed in
/usr/local/lib/ruby/gems/3.2.0+2/gems/sqlite3-1.5.2 for inspection.
Results logged to
/usr/local/lib/ruby/gems/3.2.0+2/extensions/x86_64-linux/3.2.0+2/sqlite3-1.5.2/gem_make.out
 
  /usr/local/lib/ruby/3.2.0+2/rubygems/ext/builder.rb:102:in `run'
  /usr/local/lib/ruby/3.2.0+2/rubygems/ext/ext_conf_builder.rb:28:in `build'
  /usr/local/lib/ruby/3.2.0+2/rubygems/ext/builder.rb:171:in `build_extension'
/usr/local/lib/ruby/3.2.0+2/rubygems/ext/builder.rb:205:in `block in
build_extensions'
  /usr/local/lib/ruby/3.2.0+2/rubygems/ext/builder.rb:202:in `each'
  /usr/local/lib/ruby/3.2.0+2/rubygems/ext/builder.rb:202:in `build_extensions'
  /usr/local/lib/ruby/3.2.0+2/rubygems/installer.rb:843:in `build_extensions'
/usr/local/lib/ruby/3.2.0+2/bundler/rubygems_gem_installer.rb:72:in
`build_extensions'
  /usr/local/lib/ruby/3.2.0+2/bundler/rubygems_gem_installer.rb:28:in `install'
  /usr/local/lib/ruby/3.2.0+2/bundler/source/rubygems.rb:202:in `install'
  /usr/local/lib/ruby/3.2.0+2/bundler/installer/gem_installer.rb:54:in `install'
/usr/local/lib/ruby/3.2.0+2/bundler/installer/gem_installer.rb:16:in
`install_from_spec'
/usr/local/lib/ruby/3.2.0+2/bundler/installer/parallel_installer.rb:186:in
`do_install'
/usr/local/lib/ruby/3.2.0+2/bundler/installer/parallel_installer.rb:177:in
`block in worker_pool'
  /usr/local/lib/ruby/3.2.0+2/bundler/worker.rb:62:in `apply_func'
  /usr/local/lib/ruby/3.2.0+2/bundler/worker.rb:57:in `block in process_queue'
  /usr/local/lib/ruby/3.2.0+2/bundler/worker.rb:54:in `loop'
  /usr/local/lib/ruby/3.2.0+2/bundler/worker.rb:54:in `process_queue'
/usr/local/lib/ruby/3.2.0+2/bundler/worker.rb:91:in `block (2 levels) in
create_threads'
 
An error occurred while installing sqlite3 (1.5.2), and Bundler cannot continue.
 
In Gemfile:
  sqlite3
```

Which actually isn't too bad, but I guess I've been somehow trained to not even look at that error anymore and directly reach to the log file. 

My apologies.

**Comment by simi** [comment]
@casperisfine it is still hiding the info from log actually (even providing more friendly error).

```
root@e4eb394ab146:/# cat /usr/local/lib/ruby/gems/3.2.0+2/extensions/x86_64-linux/3.2.0+2/sqlite3-1.5.2/mkmf.log
package configuration for /usr/local/lib/ruby/gems/3.2.0+2/gems/sqlite3-1.5.2/ports/x86_64-linux-gnu/sqlite3/3.39.4/lib/pkgconfig/sqlite3.pc is not found
```

What about to print out the log for non-interactive environments as part of the error and keep it the same (stored in file) for interactive ones? That way you can see CI logs with all the details, but user will be facing the same output as today. Would that work for you @voxik @casperisfine?

**Comment by casperisfine** [comment]
> Would that work for you @voxik @casperisfine?

That would work for me, even though I'm always a bit suspicious of checking `$stderr.tty?` as some CI system tend to create a TTY to get colors and such. But worst case that means no change.

**Comment by huix11** [comment]
good
